src/main/java/software/amazon/smithy/lsp/ServerState.java

```diff
@@ -0,0 +1,236 @@
+ *                    files which may be updated by didChange
+ * @param lifecycleManager Container for ongoing tasks
+ */
+public record ServerState(
+        Map<String, Project> detachedProjects,
+        Map<String, Project> attachedProjects,
+        Set<Path> workspacePaths,
+        Set<String> managedUris,
+        DocumentLifecycleManager lifecycleManager
+) {
+    private static final Logger LOGGER = Logger.getLogger(ServerState.class.getName());
+
+    /**
+     * Create a new, empty server state.
+     */
+    public ServerState() {
+        this(
+                new HashMap<>(),
+                new HashMap<>(),
+                new HashSet<>(),
+                new HashSet<>(),
+                new DocumentLifecycleManager());
+    }
+
+    /**
+     * @param uri Uri of the document to get
+     * @return The document if found and it is managed, otherwise {@code null}
+     */
+    public Document getManagedDocument(String uri) {
+        if (managedUris.contains(uri)) {
+            ProjectAndFile projectAndFile = findProjectAndFile(uri);
+            if (projectAndFile != null) {
+                return projectAndFile.file().document();
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param path The path of the document to get
+     * @return The document if found and it is managed, otherwise {@code null}
+     */
+    public Document getManagedDocument(Path path) {
+        if (managedUris.isEmpty()) {
+            return null;
+        }
+
+        String uri = LspAdapter.toUri(path.toString());
+        return getManagedDocument(uri);
+    }
+
+    ProjectAndFile findProjectAndFile(String uri) {
+        String path = LspAdapter.toPath(uri);
+        // TODO: From isDetached
+        for (Project project : attachedProjects.values()) {
+            ProjectFile projectFile = project.getProjectFile(path);
+            if (projectFile != null) {
+                detachedProjects.remove(uri);
+
+                return new ProjectAndFile(project, projectFile);
+            }
+        }
+
+        Project detachedProject = detachedProjects.get(uri);
+        if (detachedProject != null) {
+            ProjectFile projectFile = detachedProject.getProjectFile(path);
+            if (projectFile != null) {
+                return new ProjectAndFile(detachedProject, projectFile);
+            }
+        }
+
+        LOGGER.warning(() -> "Tried to unknown file: " + uri);
+
+        return null;
+    }
+
+    boolean isDetached(String uri) {
+        // We might be in a state where a file was added to a tracked project,
+        // but was opened before the project loaded. This would result in it
+        // being placed in a detachedProjects project. Removing it here is basically
+        // like removing it lazily, although it does feel a little hacky.
+        String path = LspAdapter.toPath(uri);
+        if (detachedProjects.containsKey(uri)) {
```

Comment:
nit: I think maybe we could invert this to return early.

Reply:
I refactored this a bit, and now I like it within the if since it's more clear (to me at least) that we need detachedProjects to contain the uri.